test/general/backend_inclusion.pass.cpp

```diff
@@ -0,0 +1,69 @@
+// -*- C++ -*-
+//===-- backend_inclusion.pass.cpp ----------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include _PSTL_TEST_HEADER(execution)
+
+#include "support/utils.h"
+
+// The following test suites verify that the unused backend is disabled.
+#if !_ONEDPL_PAR_BACKEND_SERIAL
+#   ifdef _ONEDPL_PARALLEL_BACKEND_SERIAL_H
+#        error The parallel serial backend is used while it should not (_ONEDPL_PAR_BACKEND_SERIAL==0)
+#   endif
+#endif
+
+#if defined(ONEDPL_USE_OPENMP_BACKEND) && !ONEDPL_USE_OPENMP_BACKEND
+#   ifdef _ONEDPL_PARALLEL_BACKEND_OMP_H
+#       error The parallel OpenMP backend is used while it should not (ONEDPL_USE_OPENMP_BACKEND==0)
+#   endif
+#endif
+
+#if defined(ONEDPL_USE_TBB_BACKEND) && !ONEDPL_USE_TBB_BACKEND
+#   ifdef _ONEDPL_PARALLEL_BACKEND_TBB_H
+#       error The parallel TBB backend is used while it should not (ONEDPL_USE_TBB_BACKEND==0)
+#   endif
+#endif
+
+#if defined(ONEDPL_USE_DPCPP_BACKEND) && !ONEDPL_USE_DPCPP_BACKEND
+#   ifdef _ONEDPL_parallel_backend_sycl_H
+#       error The parallel DPC++ backend is used while it should not (ONEDPL_USE_DPCPP_BACKEND==0)
+#   endif
+#   ifdef _ONEDPL_parallel_backend_sycl_fpga_H
+#       error The DPC++ backend for the FPGA is used while it should not (ONEDPL_USE_DPCPP_BACKEND==0)
+#   endif
+#endif
+
+#if !ONEDPL_FPGA_DEVICE
+#   ifdef _ONEDPL_parallel_backend_sycl_fpga_H
+#       error The DPC++ backend for the FPGA is used while it should not (ONEDPL_FPGA_DEVICE==0)
+#   endif
+#endif
+
+// The following test suites verify that the required backend is enabled.
+#if defined(ONEDPL_USE_TBB_BACKEND) && ONEDPL_USE_TBB_BACKEND
+#   if !defined(_ONEDPL_PARALLEL_BACKEND_TBB_H)
+#       error The parallel TBB backend is not used while it should (ONEDPL_USE_TBB_BACKEND==1)
+#   endif
+#elif defined(ONEDPL_USE_OPENMP_BACKEND) && ONEDPL_USE_OPENMP_BACKEND
+#   if !defined(_ONEDPL_PARALLEL_BACKEND_OMP_H)
+#       error The parallel OpenMP backend is not used while it should (ONEDPL_USE_OPENMP_BACKEND==1)
+#   endif
+#endif

```

Comment:
@rarutyun, what is this test intended to check?

It does not cover the full table of TBB/OpenMP/serial selection, only the part where either TBB, OpenMP, or both are explicitly requested. If not that "or both" part that is covered by the check order and `elif`, it looks pretty similar to the first set of tests above, with the condition changed to "explicitly enabled". If that's the intent, why the same is not done for DPC++ backends?

Reply:
It intended to check the public API when it's the backend is explicitly requested/disabled by the user. As you said it only checks the case with explicit macro definitions and skips some cases for the "Undefined" part. 
I believe that it deserves to be a separate test because TBB and OpenMP backends are mutually exclusive. The test basically checks that backend priority logic works as expected. Besides "both" scenario it also covers the case that when TBB backend is disabled or undefined the priority is given to OpenMP backend.

Not sure if we need such a test for DPC++ part. We can add such but I don't think it brings much value. The initial intent of this test was to check if we don't include something we should not. For DPC++ it's already covered. If something DPC++ specific (we use in the tests) is missed we got compile time error. Thus, I tend to think it's not so important to check those.

As I said additional test for TBB and OpenMP is here because or backend selection logic and mutual exclusiveness.

Reply:
OK, I see the point, but as the test of backend selection logic and mutual exclusiveness, it's quite incomplete. Yes it checks the backend priority  - but only if both are defined, and neither if both are undefined nor if both are disabled. And it does not check mutual exclusiveness at all - for that, you'd need to check that only one of the three backends is enabled under any given condition.

> Besides "both" scenario it also covers the case that when TBB backend is disabled or undefined the priority is given to OpenMP backend.

More accurately, when TBB macro is undefined and OpenMP macro is defined, the priority is given to OpenMP (the "TBB disabled" is not interesting from the priority viewpoint). I agree with that, thanks.

Reply:
Just to clarify, I do not insist this test to cover all possible macro combinations; it's OK to miss some that are not very important or would be complicated to check properly. I do not even request that this test should exist. But if it does exist, I think for now it's not sufficient to match the stated goal.